### PR TITLE
At 13 File Uploader

### PR DESCRIPTION
**Pull request by OllieVanD**
Created file uploader with routers and 4 classes to handle csv file with validation checks Unit tested with jasmine and feature tested with cypress.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module.exports = {
```
Comment (dearshrewdwit) [inline .eslintrc.js:1]: Nice idea. Let's move this to its own PR, however.

File: cypress/support/commands.js [hunk 2]
```diff
@@ -23,3 +23,17 @@
 //
 // -- This will overwrite an existing command --
 // Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+
+
+
+Cypress.Commands.add('form_request', ( method, url, formData, done) => {
```
Comment (dearshrewdwit) [inline cypress/support/commands.js:29]: What's the reason for this command? Where is it used?

File: package.json [hunk 3]
```diff
@@ -11,15 +11,24 @@
...
     "express": "^4.17.1",
+    "express-fileupload": "^1.2.1",
     "jasmine": "^3.8.0",
     "mochawesome": "^6.2.2",
     "nodemon": "^2.0.12",
     "pg": "^8.7.1",
     "pg-hstore": "^2.3.4",
     "sequelize": "^6.6.5",
     "sequelize-cli": "^6.2.0"
+  },
```
Comment (dearshrewdwit) [inline package.json:26]: These should be added in a separate PR with eslint

File: package.json [hunk 4]
```diff
@@ -11,15 +11,24 @@
   "license": "ISC",
   "dependencies": {
     "cors": "^2.8.5",
+    "csv-parser": "^3.0.0",
```
Comment (dearshrewdwit) [inline package.json:14]: Where is this dependency used?

File: src/fileUpload.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    if (assessmentClass === undefined) {
+      return 'invalid table selected'
+    } else return assessmentClass
+  }
+
+  async dataChecks () {
+    if (this.table === ModuleChallenge) {
+      const moduleCheck = new ModuleChallengeChecker()
+      this.errors = await moduleCheck.check(this.data)
+      return this.errors
```
Comment (dearshrewdwit) [inline src/fileUpload.js:50]: What's the reasoning behind  explicitly returning from this function? I don't see it used in the controller action

File: src/fileUpload.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }
+  }
+
+  async addToDatabase () {
+    if (this.errors.length === 0) {
+      await this.table.bulkCreate(this.data, {
+        returning: true
+      })
+      this.status = 'Success'
```
Comment (dearshrewdwit) [inline src/fileUpload.js:63]: What's the reasoning behind setting the status property? I don't see it used in the response object

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]
Good work so far.

Move the eslint addition to its own PR and resolve some of the questions below in comments. Additionally

Write some test cases that cover:

- What would happen uploading a csv with malformed headers? No headers?
- What happens if one of the text fields for the learning self assessment is beyond the dB constraint for character length?

File: src/fileUpload.js (hunk 6 above)
Comment (OllieVanD) [reply src/fileUpload.js:62]: This is something that we use for AT-34 branch, must have just been put on this branch by accident.

File: cypress/support/commands.js (hunk 2 above)
Comment (OllieVanD) [reply cypress/support/commands.js:29]: Fixed

File: package.json (hunk 3 above)
Comment (OllieVanD) [reply package.json:23]: Can you elaborate with what you mean by this?

Comment (dearshrewdwit) [reply package.json:23]: Hm, cheeky, you rewrote history, so what can I say!

Originally, this was a comment on the additional `eslint` node packages that appeared here.

Comment (OllieVanD) [reply package.json:23]: I have added additional eslint node packages to the devdependencies, should they not be in the project?

File: README.md [hunk 7]
```diff
@@ -25,5 +25,8 @@ $ npx cypress run --reporter mochawesome --reporter-options "reportDir=cypress/r
 # to run unit tests
 $ npm test
 ```
+<<<<<<< HEAD
```
Comment (dearshrewdwit) [inline README.md:28]: Still a couple remnants from an attempted conflict resolution to remove.

File: src/fileUpload.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+  checkArraysAreEqual(arr1, arr2) {
+    if (arr1.length !== arr2.length) return false 
+    for (let i = 0; i < arr1.length; i++) {
+      if (arr1[i] !== arr2[i]) return false 
+    }
+    return true
+  }
+
+  headerChecker(headers) {
```
Comment (dearshrewdwit) [inline src/fileUpload.js:23]: We'll come back to this to break this apart.

**Review by dearshrewdwit — APPROVED** [review]
Ok. There's a bunch more work needed to refactor, but let's get this in now.

**Review by dearshrewdwit — COMMENTED** [review]
See comment re: conflict resolution

**Review by Neokrieger — APPROVED** [review]
LGTM

**Review by dearshrewdwit — APPROVED** [review]

**Pull request merged**